Promise all

king-11 opened the pull request:
<!--- Tip: You don't have to remove these comments -->

## Description
<!--- Describe your changes in detail -->
Update sequentially running loops for promise resolution to concurrently running promises using Promise.all while updating tests for two functions as promises are now run concurrently.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
promise resolution sequentially takes a lot of time to resolve all when then can instead run concurrently.
<!--- If it fixes an open issue, please link to the issue here. -->
closes #287 

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have updated the documentation accordingly.
- [ ] I have added new practice to practice list in README.md.
- [x] I have read the **CONTRIBUTING** document.
- [x] I haven't repeated the code. (DRY)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

adelkahomolova commented:
Hi, @king-11 there is one test failing. Could you look at it, please? If you'll need any help, let me now. It seems it's just the wrong use of nock (which is not hard to use it badly actually...).

<img width="1058" alt="Screenshot 2020-10-05 at 09 35 58" src="https://user-images.githubusercontent.com/53510747/95051933-5b610480-06ee-11eb-87b8-11d53f6a2167.png">

king-11 commented:
is HTTP the expected error? then I shall make amends in the test to check for it

adelkahomolova commented:
No, this error is not expected. It tells us just the fact that the nock request does not correspond with the request which is made in `git.flatTraverse()` method. Maybe some of the calls are not executed. 🤔  @king-11

king-11 commented:
@adelkahomolova i was able to trace back the issue to ```src/services/git/GitHubService.ts:332:41``` on line
```js
try {
        response = <GetContentsResponse>await this.unwrap(this.client.repos.getContent({ owner, repo, path }));
      } catch (e) {
        if (e.name !== 'HttpError' || e.status !== 404) {
          throw e;
        }

        return null;
      }
```
but am not able to further debug

adelkahomolova commented:
Hey, you shouldn't change the code just because of the test.
Please, keep your code in `Git.ts` and change the test. I think that just remove the calling second nock should help. It should fail fast in this case. I can explain it more, why we don't want to use Promise.allSettled() if you will need it or want it :) @king-11

king-11 commented:
@adelkahomolova ahh sorry :sweat_smile:  I am not much aware abt nock calls so i thought this can help but i will try as you say :+1: and Yeah, I would also like to hear why Promise.allSettled isn't correct :

adelkahomolova commented:
The problem is, that you reject the whole array of promises after all, even if there is just one call that you rejected. If you will use `Promise.allSettled` it will make a call for every single item in the array, even if it hits a snag. Then you will take a whole array and check if one of the items was rejected. If yes you'll reject the whole array. So it takes a lot of time. But if you will use the `Promise.all()`, it is rejected right after getting the first rejection.
I hope I explain it understandable 😄 . Ask me, if you have any questions. :) @king-11

king-11 commented:
@adelkahomolova  yeah understood this that time is an issue while using allSettled. but as you said 
so i changed test as follows and its passing now does it look good to you ?

```js
it('stops on false', async () => {
      gitHubNock.getDirectory('mockFolder', ['mockFile.ts'], ['mockSubFolder']);
      gitHubNock.getFile('mockFolder/mockFile.ts');
      gitHubNock.getDirectory('mockFolder/mockSubFolder', ['mockSubFolderFile.txt'], []);
      gitHubNock.getFile('mockFolder/mockSubFolder/mockSubFolderFile.txt');

      let response = true;
      await git
        .flatTraverse('mockFolder', () => {
          return false;
        })
        .catch((e) => (response = e));

      expect(response).toEqual(false);
    });
```

adelkahomolova commented:
<img width="804" alt="Screenshot 2020-10-05 at 16 05 38" src="https://user-images.githubusercontent.com/53510747/95090000-14423600-0725-11eb-931f-78b545a472b3.png">

Try something like this so we use [DRY principle](https://cs.wikipedia.org/wiki/Don%27t_repeat_yourself). @king-11 Thanks!

king-11 commented:
@adelkahomolova DRY is an issue but while I am using gitHubNock.getDirectory() it should fetch all files and dirs but it isn't doing that for some reason :thinking:

king-11 commented:
for dry principle, I think we can use before all and have a nested describe inside flat traverse which can use beforeAll and instantiate the directories becuase there are two tests that need complete instantiation

prokopsimek commented:
@king-11 What about to change the catch to:

```ts
it('stops on false', async () => {
        let response = true;
        await git
          .flatTraverse('mockFolder', () => {
            return false;
          })
          .catch((e) => (fail(response)));

        expect(response).toEqual(false);
      });
```
?

I am fine with your solution. We could fail the test with a relevant error message with jest's `fail` function.

king-11 commented:
@prokopsimek ohhh i didn't know abt that we check for fails as well 🤔 that will be great i believe i will look into it and made necessary amends.

adelkahomolova commented:
I think @prokopsimek thinks that your solution is good. He just wanted to fail the test with a relevant message, so you should just change `(response = e)` for `(fail(response))`. @king-11

king-11 commented:
@adelkahomolova catch is the place through which am getting value for a response as false. e is actually just a false value which I am assigning to the response which is then checked in the expect function if I change that then the response won't get a false value and will fail even though e returned false :sweat_smile: so I am not able to get the use of fail here can you please tell me a bit more about its use case here

king-11 commented:
@adelkahomolova still to push the changes that i cited just a sec :sweat_smile:

adelkahomolova commented:
Great, this should work. So when you'll implement it, I'll merge it. 🎉  
I really appreciate the time you invest in the dx-scanner. I hope you enjoyed contributing to our repo! I'll be glad if you'll let us some feedback either for communication and for the dx-scanner itself. Every thought is valuable for us. Thanks, @king-11

king-11 commented:
rebased with upstream and added changes and @adelkahomolova  @prokopsimek  thanks a lot really for helping out throughout this is my first major contribution and thanks to you guys I received a lot of help and really enjoyed contributing to dx-scanner will continue doing that as its a really great tool. and run failed :'( api rate limit again

adelkahomolova commented:
I'm glad to hear that! ❤️  
I know, the API rate limit is the issue we're trying to solve. Sorry for that 😢

prokopsimek commented:
:tada: This PR is included in version 3.40.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/dx-scanner/v/3.40.0)
- [GitHub release](https://github.com/DXHeroes/dx-scanner/releases/tag/v3.40.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
